### ISSUE TITLE
Gradient of squared 2-norm returns NaN when input is vector of zeros

### ISSUE THREAD

**Issue by sdwill** [google/jax#4177]
The squared l2 norm of a vector x, ||x||^2, is the sum of the squares (squared magnitudes if the vector is complex-valued) of the elements.  The gradient is simply 2*x.  In Jax, providing a vector of zeros to the gradient function unexpectedly results in NaN output.  Here's a minimal example to demonstrate, with the trivial sum-of-squares implementation for comparison:
```python
import jax.numpy as np
from jax import grad


vec = np.zeros(2)

def squared_2_norm(x):
    return np.linalg.norm(x) ** 2

def sum_of_squares(x):
    return np.sum(x ** 2)

gradient_norm = grad(squared_2_norm)
gradient_sum = grad(sum_of_squares)

print(squared_2_norm(vec))
print(sum_of_squares(vec))
print(gradient_norm(vec))
print(gradient_sum(vec))
```

Running this produces the following output:
```
0.0
0.0
[nan nan]
[0. 0.]
```

Interestingly, when `vec` is altered to `vec = np.ones(2)`, the results are almost identical, but the version using `np.norm()` seems to have some floating-point error:
```
1.9999999
2.0
[1.9999999 1.9999999]
[2. 2.]
```

**Comment by jakevdp** [comment]
I think the issue here is that the gradient of `norm(x)` is not well-defined at zero. It's similar to the gradient of abs issue discussed in #10515

I'm going to close this as I think it's working as intended.
